src/wasm/WasmModule.cpp

```diff
@@ -430,69 +426,77 @@ std::future<int32_t> WasmModule::executeOpenMPTask(threads::OpenMPTask t)
+    int threadPoolIdx =
+      (t.nextLevel->depth - 1 + t.msg->ompthreadnum()) % threadPoolSize;
+
+    // Enqueue the task
+    openMPTaskQueueMap[threadPoolIdx].enqueue(
+      std::make_pair(std::move(p), std::move(t)));
+
+    // Check if we need to add the thread for this index
+    if (openMPThreads.count(threadPoolIdx) == 0) {
+        // Get mutex and re-check
         faabric::util::UniqueLock lock(threadsMutex);
 
-        // Double check once lock acquired
-        if (openMPThreads.size() < threadPoolSize) {
+        if (openMPThreads.count(threadPoolIdx) == 0) {
             if (threadStacks.empty()) {
                 throw std::runtime_error("Run out of thread stacks");
             }
 
-            int32_t threadPoolId = openMPThreads.size();
-
             uint32_t stackTop = threadStacks.back();
             threadStacks.pop_back();
 
-            openMPThreads.emplace_back([this, stackTop, threadPoolId] {
-                int queueTimeoutMs =
-                  faabric::util::getSystemConfig().boundTimeout;
-                auto logger = faabric::util::getLogger();
-                logger->debug("Starting OpenMP pool thread {} (timeout {}ms)",
-                              threadPoolId,
-                              queueTimeoutMs);
-
-                for (;;) {
-                    std::pair<std::promise<int32_t>, threads::OpenMPTask>
-                      taskPair = openMPTaskQueue.dequeue(queueTimeoutMs);
-
-                    if (taskPair.second.isShutdown) {
-                        taskPair.first.set_value(0);
-
-                        logger->debug(
-                          "OpenMP thread pool thread {} shutting down",
-                          threadPoolId);
-
-                        break;
-                    }
-
-                    logger->debug("OpenMP {}: executing OMP thread {}, "
-                                  "function {}, message {}",
-                                  threadPoolId,
-                                  taskPair.second.msg->ompthreadnum(),
-                                  taskPair.second.msg->funcptr(),
-                                  taskPair.second.msg->id());
-
-                    auto& sch = faabric::scheduler::getScheduler();
-
-                    // We are now in a new thread so need to set up
-                    // everything that uses TLS
-                    setUpOpenMPContext(taskPair.second.threadIdx,
-                                       taskPair.second.nextLevel);
-                    setExecutingModule(this);
-                    setExecutingCall(taskPair.second.parentMsg);
-
-                    taskPair.first.set_value(
-                      executeAsOMPThread(stackTop, taskPair.second.msg));
-
-                    // Caller has to notify scheduler when finished
-                    // executing a thread locally
-                    sch.notifyCallFinished(*taskPair.second.msg);
-                }
-            });
+            openMPThreads.insert(
```

Comment:
Mostly just indentation here

Reply:
Is there a reason for switching from `emplace`-ing to `insert`-ing? I would personally always go for the former as it can sometimes be benefficial, and won't harm.